src/useMultiRootEditor.tsx

```diff
@@ -0,0 +1,524 @@
+			// could be fired by <CKEditorContext /> and <CKEditor /> at the same time, this `setTimeout()` makes sure
+			// that <CKEditorContext /> component will be destroyed first, so during the code execution
+			// the `ContextWatchdog#state` would have a correct value. See `EditorWatchdogAdapter#destroy()` for more information.
+			/* istanbul ignore next */
+			setTimeout( async () => {
+				if ( watchdog ) {
+					await watchdog.destroy();
+					watchdog = null;
+
+					return resolve();
+				}
+
+				if ( editor ) {
+					await editor.destroy();
+
+					return resolve();
+				}
+
+				resolve();
+			} );
+		} );
+	};
+
+	/**
+	 * Initializes the editor by creating a proper watchdog and initializing it with the editor's configuration.
+	 */
+	const _initializeEditor = async (): Promise<void> => {
+		if ( props.disableWatchdog ) {
+			await _createEditor( props.content, _getConfig() );
+
+			return;
+		}
+
+		/* istanbul ignore next */
+		if ( watchdog ) {
+			return;
+		}
+
+		if ( context instanceof ContextWatchdog ) {
+			watchdog = new EditorWatchdogAdapter( context );
+		} else {
+			watchdog = new EditorWatchdog( props.editor, props.watchdogConfig );
+		}
+
+		watchdog.setCreator( ( data, config ) => _createEditor( data as Record<string, string>, config ) );
+
+		watchdog.on( 'error', ( _, { error, causesRestart } ) => {
+			const onError = props.onError || console.error;
+
+			onError( error, { phase: 'runtime', willEditorRestart: causesRestart } );
+		} );
+
+		await watchdog
+			.create( content as any, _getConfig() )
+			.catch( error => {
+				const onError = props.onError || console.error;
+
+				onError( error, { phase: 'initialization', willEditorRestart: false } );
+			} );
+	};
+
+	useEffect( () => {
+		if ( !editor ) {
+			return;
+		}
+
+		if ( !shouldUpdateEditor.current ) {
+			shouldUpdateEditor.current = true;
+
+			return;
+		}
+
+		const editorData = editor.getFullData();
+		const editorAttributes = editor.getRootsAttributes();
+
+		const {
+			addedKeys: newRoots,
+			removedKeys: removedRoots,
+			modifiedKeys: modifiedRoots
+		} = _getStateDiff( editorData, content || {} );
+
+		const {
+			addedKeys: newAttributes,
+			removedKeys: removedAttributes,
+			modifiedKeys: modifiedAttributes
+		} = _getStateDiff( editorAttributes, attributes || {} );
+
+		const newRootAttributes = newAttributes.filter( rootName =>
+			!newRoots.includes( rootName ) && attributes[ rootName ] !== undefined );
+		const removedRootAttributes = removedAttributes.filter( rootName =>
+			!removedRoots.includes( rootName ) && attributes[ rootName ] === undefined );
+		const modifiedRootAttributes = modifiedAttributes.filter( rootName => attributes[ rootName ] !== undefined );
+
+		editor.model.change( writer => {
+			_handleRoots( newRoots, removedRoots, modifiedRoots );
+			_handleRootsAttributes( writer, newRootAttributes, removedRootAttributes, modifiedRootAttributes );
+		} );
```

Comment:
I don't like this part of the integration.

First, every time I get here, I have to re-understand which values mean what. Lack of comments and naming of the variables do not help.

Second, it seems to me that any state where `content` and `attributes` have different keys (different roots) is incorrect. We could check this and throw if that happens.

Third, if a root was added or removed, AFAIR, you don't need to separately add attributes, because you already adding them in `editor.addRoot()` (line 447).

Given that, it doesn't make sense to separately evaluate `newRoots`, `removedRoots`, `newAttributes` and `removedAttributes`. Instead, we should simplify this because there's so many diffs and filtering going on that it is hard to wrap your head about it:

1.  Evaluate which roots were added and which were removed.
2.  Add new roots with attributes.
3.  Detach removed roots.
4.  For the rest, check which changed data. If any, `editor.data.set()`.
5.  As you check if roots data was changed, you can also check if roots attributes changed. If so, re-set the attributes.
    1.  In general, I don't like clear all + set all approach due to extra operations created, but it should not matter much. It shouldn't be many attributes (operations), and it shouldn't trigger any extra actions because differ will "squash" the changes. So I am fine keeping it like this.

Add some inline comments and you will be also able to name variables better as it will be less of them and they will have more precise meaning.